scripts/helper_scripts/unipept-database-rs/src/calculate_lcas/taxonomy.rs

```diff
@@ -0,0 +1,161 @@
+    }
+
+    pub fn calculate_lcas(&self) -> Result<()> {
+        let reader = open_sin();
+
+        let mut current_sequence = String::new();
+        let mut taxa: Vec<i32> = Vec::new();
+
+        for (i, line) in reader.lines().enumerate() {
+            if i % 10000000 == 0 && i != 0 {
+                eprintln!("{}: {}", now_str(), i);
+            }
+
+            let line = line.context("error reading line from stdin")?;
+
+            let (sequence, taxon_id) =
+                line.split_once(SEPARATOR).context("error splitting line")?;
+            let taxon_id: i32 = taxon_id
+                .trim_end()
+                .parse()
+                .context("error parsing taxon id to int")?;
+
+            if current_sequence.is_empty() || current_sequence != sequence {
+                if !current_sequence.is_empty() {
+                    self.handle_lca(&current_sequence, self.calculate_lca(&taxa));
+                }
+
+                current_sequence = sequence.to_string();
+                taxa.clear();
+            }
+
+            taxa.push(taxon_id);
+        }
+
+        self.handle_lca(&current_sequence, self.calculate_lca(&taxa));
+        Ok(())
+    }
+
+    fn calculate_lca(&self, taxa: &[i32]) -> i32 {
+        let mut lca = 1;
+
+        let lineages: Vec<&Vec<i32>> = taxa
+            .iter()
+            .map(|x| &self.taxonomy[*x as usize])
+            .filter(|x| !x.is_empty())
+            .collect();
+
+        for rank in 0..RANKS {
+            let final_rank = rank;
+            let mut value = -1;
+
+            let iterator = lineages
+                .iter()
+                .map(|&x| x[final_rank as usize])
+                .filter(|&x| {
+                    if final_rank == GENUS || final_rank == SPECIES {
+                        x > 0
+                    } else {
+                        x >= 0
+                    }
+                });
+
+            let mut all_match = true;
+
+            // This was near-impossible to do with the iterators above,
+            // so we're using a simplified loop here
+            for item in iterator {
+                if value == -1 {
+                    value = item;
+                } else if item != value {
+                    all_match = false;
+                    break;
+                }
+            }
+
+            // If we found a new value that matched for all of them, use this as the new best
+            if value != -1 {
+                // If not everything matched, this is not a common ancestor anymore,
+                // so we can stop
+                if !all_match {
+                    break;
+                }
+
+                if value != 0 {
+                    lca = value;
+                }
+            }
+        }
+
+        lca
+    }
+
+    fn handle_lca(&self, sequence: &String, lca: i32) {
+        println!("{}\t{}", sequence, lca);
+    }
+}
+
+fn parse_int(s: &String) -> Result<i32> {
+    if s == NULL {
```

Comment:
I was really confused by this, thinking Rust made the billion dollar mistake after all...

I suggest changing `NULL` to `NULL_FIELD` or `NULL_STRING`, or adding a comment that `NULL` is here not in fact an actual null value.